agent.go

```diff
@@ -54,6 +56,9 @@ type sandbox struct {
 	grpcListener net.Listener
 	sharedPidNs  bool
 	mounts       []string
+
+	procChLock  sync.RWMutex
+	processesCh map[int]chan int
```

Comment:
the exitcode should be saved on type process.
the type process for the exec-processes will be removed when a successful WaiProcess()
the type process for the init-process will be removed when RemoveContainer()

Reply:
Not sure I follow here. Is there something wrong with the way I have implemented the channel message delivery between the subreaping loop and the WaitProcess() ?

Reply:
There is nothing incorrect with the processesCh.
I hope the variables for the process are put into the type process.

Reply:
Ok I get it, but the issue by trying to find the right process entry is that we're going to loop over every container map of processes and check for each one if there is a corresponding entry. Having this very simple map at the pod level simplifies things and we don't need all the informations contained into a `process` structure.

Reply:
How about adding `map[int] process` in the `sandbox`?

Reply:
type ProcessID int

Reply:
Oh you mean somehting like this ?
```go
type ProcessID int

type sandbox struct {
        ...
        processesCh map[int]chan ProcessID
```

I need this map to be a map of channels since I expect WaitProcess() to listen on this channel until it gets the exit code.

Reply:
add `map[ProcessID] process` to sandbox
add `chan exitcode` to type process

Reply:
Ok this would work but this means we would have a map of process structures and we would only use the exit code. That's a bit too much, don't you think ?

I could add map[ProcessID] ExitCode with both ProcessID and ExitCode being int types, this would be simpler.

Reply:
I don't want the fields of a process to be scattered.
If it is complicated due to the locking, the locking can be simplified by only using the lock in the sandbox and removing the lock in container.

Reply:
Well this field is not scattered since it's not even needed into the `process` structure.

Reply:
@laijs It's more about duplication than locking complexity. Each `container` already has a `processes   map[int]*process` map and we'd be duplicating this information in the sandbox now. And we'd have to keep those maps coherent, which adds complexity as well.
If we want to keep all processes information in a single structure (Which I agree would make sense), I'd prefer to keep one PID <-> process map per container and no more. We could maintain a PID <-> container id map at the sandbox level to avoid going through all containers to find a process when being given a PID

Reply:
@laijs any comment ?

Reply:
The death messages of the other helper processes created directly by agent are routed to this loop. We need a mechanism to help programing and to replace exec.Cmd.Wait() or os.Process.Wait(). We can reuse the mechanism in containerd or implement a new one.

Reply:
@laijs I am sorry I really don't follow on this one. Can you try to explain what you need me to implement and why ? Also, I don't get how the last comment is related to the entire conversation.

Reply:
@sboeuf when agent is subreaper, exec.Cmd.Wait() and os.Process.Wait() for helper processes are not workable, we should provide a replacement.

Reply:
@sboeuf it is quite simple in containerd.
https://github.com/containerd/containerd/blob/master/reaper/reaper.go

Reply:
@laijs What do you mean by "helper processes" ?
Also, do you expect that we provide a different implementation of `Wait()` so that we can override the standard behavior ? How can we do that exactly ?
I have looked at the containerd code and I think it partially does what we do, or am I missing something that you expect me to add to our code ?

Reply:
@laijs actually I get it now :)
You expect that we define a new way to spawn processes in case we want to use something simple like `exec.Cmd.Wait()`. I am sorry it took me so long but I thought you were trying to solve an issue in the current implementation. Instead you're anticipating that we won't be able to rely on the stardard `exec` and `os` functions since we would end up in a racy case where we would not know who will get to reap the child process first (between this function call and our subreaper).
I'll implement this and update the PR.